File: .github/scripts/test-browser-interface-profile.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "fileutils"
require "find"
require "open3"
require "rbconfig"
require "tempfile"
require "timeout"
require "tmpdir"

CommandResult = Struct.new(:stdout, :stderr, :status, :timed_out, keyword_init: true)

fixture_root = File.expand_path("../fixtures/profiles/browser-interface", __dir__)
validator = File.expand_path("validate-skill-repository.rb", __dir__)
expected_files = %w[
  Gemfile
  RUNTIME.md
  SKILL.md
  WEB_INTERFACE.md
  src/text_stats.rb
  tests/test_web_server.rb
  web/server.rb
].sort.freeze

terminate_and_wait = lambda do |pid|
  begin
    Process.kill("TERM", pid)
  rescue Errno::ESRCH
    return nil
  end

  begin
    Timeout.timeout(1) do
      _waited_pid, status = Process.wait2(pid)
      return status
    end
  rescue Timeout::Error
    begin
      Process.kill("KILL", pid)
    rescue Errno::ESRCH
      nil
    end
    _waited_pid, status = Process.wait2(pid)
    status
  rescue Errno::ECHILD
    nil
  end
end

run_command = lambda do |*command, chdir:, timeout_seconds:, env: {}|
  stdout_file = Tempfile.new("browser-fixture-stdout")
  stderr_file = Tempfile.new("browser-fixture-stderr")
  status = nil
  timed_out = false
  begin
    pid = Process.spawn(
      env,
      *command,
      chdir: chdir,
      in: File::NULL,
      out: stdout_file.path,
      err: stderr_file.path
    )
  rescue Errno::ENOENT => error
    return CommandResult.new(
      stdout: "",
      stderr: "#{error.message}\n",
      status: nil,
      timed_out: false
    )
  end

  begin
    Timeout.timeout(timeout_seconds) do
      _waited_pid, status = Process.wait2(pid)
    end
  rescue Timeout::Error
    timed_out = true
    status = terminate_and_wait.call(pid)
  ensure
    stdout_file.flush
    stderr_file.flush
  end

  CommandResult.new(
    stdout: File.binread(stdout_file.path),
    stderr: File.binread(stderr_file.path),
    status: status,
    timed_out: timed_out
  )
ensure
  stdout_file&.close!
  stderr_file&.close!
end

failures = []
actual_files = Find.find(fixture_root).filter_map do |path|
  next if path == fixture_root || File.directory?(path)

  path.delete_prefix("#{fixture_root}/")
end.sort

if actual_files != expected_files
  failures << "browser-interface: expected reduced layout #{expected_files.inspect}, got #{actual_files.inspect}"
end

runtime_contract = File.read(File.join(fixture_root, "RUNTIME.md"), encoding: "UTF-8")
documented_commands = [
  "bundle install",
  "TEXT_STATS_WEB_ENABLED=1 bundle exec ruby web/server.rb",
  "bundle exec ruby web/server.rb --stop",
  "bundle exec ruby web/server.rb --health",
  "bundle exec ruby tests/test_web_server.rb"
].freeze

documented_commands.each do |command|
  unless runtime_contract.include?(command)
    failures << "browser-interface runtime: missing documented command #{command.inspect}"
  end
end

bundle_check = run_command.call("bundle", "--version", chdir: fixture_root, timeout_seconds: 10)
if bundle_check.timed_out || !bundle_check.status&.success?
  failures << "browser-interface dependencies: Bundler is required by the fixture workflow; stderr=#{bundle_check.stderr.inspect}"
end

if failures.empty?
  Dir.mktmpdir("browser-interface-profile") do |directory|
    FileUtils.cp_r("#{fixture_root}/.", directory)
    run_command.call("git", "init", "--quiet", chdir: directory, timeout_seconds: 10)
    run_command.call("git", "add", ".", chdir: directory, timeout_seconds: 10)

    validation = run_command.call(
      RbConfig.ruby,
      validator,
      chdir: directory,
      timeout_seconds: 30,
      env: { "RUBYOPT" => nil }
    )
    unless validation.status&.success? && !validation.timed_out
      failures << "browser-interface: expected complete repository validation to pass; " \
                  "stdout=#{validation.stdout.inspect}, stderr=#{validation.stderr.inspect}, " \
                  "timed_out=#{validation.timed_out}"
    end

    bundle_config = run_command.call(
      "bundle",
      "config",
      "set",
      "--local",
      "path",
      ".bundle",
      chdir: directory,
      timeout_seconds: 30
    )
    unless bundle_config.status&.success? && !bundle_config.timed_out
      failures << "browser-interface dependencies: expected local Bundler path configuration to pass; " \
                  "stderr=#{bundle_config.stderr.inspect}"
    end

    bundle_install = run_command.call(
      "bundle",
      "install",
      "--jobs",
      "4",
      "--retry",
      "3",
      chdir: directory,
      timeout_seconds: 180
    )
    unless bundle_install.status&.success? && !bundle_install.timed_out
      failures << "browser-interface dependencies: expected bundle install success; " \
                  "stdout=#{bundle_install.stdout.inspect}, stderr=#{bundle_install.stderr.inspect}, " \
                  "timed_out=#{bundle_install.timed_out}"
    end

    if bundle_install.status&.success?
      %w[src/text_stats.rb web/server.rb tests/test_web_server.rb].each do |path|
        syntax = run_command.call(
          RbConfig.ruby,
          "-c",
          path,
          chdir: directory,
          timeout_seconds: 10
        )
        unless syntax.status&.success? && !syntax.timed_out
          failures << "browser-interface syntax: expected #{path} to parse; stderr=#{syntax.stderr.inspect}"
        end
      end

      tests = run_command.call(
        "bundle",
        "exec",
        RbConfig.ruby,
        "tests/test_web_server.rb",
        chdir: directory,
        timeout_seconds: 90
      )
      unless tests.status&.success? && !tests.timed_out
        failures << "browser-interface tests: expected success; stdout=#{tests.stdout.inspect}, " \
                    "stderr=#{tests.stderr.inspect}, timed_out=#{tests.timed_out}"
      end

      implementation_path = File.join(directory, "src/text_stats.rb")
      missing_path = "#{implementation_path}.missing"
      File.rename(implementation_path, missing_path)
      missing_implementation = run_command.call(
        "bundle",
        "exec",
        RbConfig.ruby,
        "web/server.rb",
        chdir: directory,
        timeout_seconds: 10,
        env: {
          "TEXT_STATS_WEB_ENABLED" => "1",
          "TEXT_STATS_WEB_PORT" => "0",
          "TEXT_STATS_WEB_PID_FILE" => File.join(directory, "missing.pid")
        }
      )
      if missing_implementation.timed_out || missing_implementation.status&.success? ||
         missing_implementation.stderr.strip.empty?
        failures << "browser-interface missing implementation: expected a prompt nonzero failure with diagnostics; " \
                    "status=#{missing_implementation.status&.exitstatus.inspect}, " \
                    "stderr=#{missing_implementation.stderr.inspect}, timed_out=#{missing_implementation.timed_out}"
      end
      File.rename(missing_path, implementation_path)
    end
  end
end

Dir.mktmpdir("invalid-browser-interface-profile") do |directory|
  FileUtils.cp_r("#{fixture_root}/.", directory)
  File.delete(File.join(directory, "WEB_INTERFACE.md"))
  run_command.call("git", "init", "--quiet", chdir: directory, timeout_seconds: 10)
  run_command.call("git", "add", ".", chdir: directory, timeout_seconds: 10)

  validation = run_command.call(
    RbConfig.ruby,
    validator,
    chdir: directory,
    timeout_seconds: 30,
    env: { "RUBYOPT" => nil }
  )
  if validation.timed_out || validation.status&.success?
    failures << "browser-interface invalid contract: expected missing WEB_INTERFACE.md to fail repository validation"
  elsif !validation.stderr.include?("WEB_INTERFACE.md")
    failures << "browser-interface invalid contract: expected an actionable WEB_INTERFACE.md diagnostic; " \
                "stderr=#{validation.stderr.inspect}"
  end
end

unless failures.empty?
  failures.each { |failure| warn failure }
  exit 1
end

puts "Browser-interface profile fixture tests passed."

File: .github/scripts/test_browser_interface_profile.py
#!/usr/bin/env python3
"""End-to-end regression harness for the browser-interface fixture."""

from __future__ import annotations

import os
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
FIXTURE = ROOT / ".github/fixtures/profiles/browser-interface"
VALIDATOR = ROOT / "template/.github/scripts/validate_skill_repository.py"
EXPECTED_FILES = sorted(
    [
        "RUNTIME.md",
        "SKILL.md",
        "WEB_INTERFACE.md",
        "public/app.css",
        "public/app.js",
        "public/index.html",
        "src/text_stats.py",
        "tests/test_web_server.py",
        "web/server.py",
    ]
)


def clean_env(extra: dict[str, str] | None = None) -> dict[str, str]:
    env = os.environ.copy()
    for key in ("PYTHONPATH", "GIT_DIR", "GIT_INDEX_FILE", "GIT_WORK_TREE"):
        env.pop(key, None)
    env["PYTHONUTF8"] = "1"
    env["PYTHONIOENCODING"] = "utf-8"
    if extra:
        env.update(extra)
    return env


def run(
    command: list[str],
    *,
    cwd: Path,
    timeout: float = 90.0,
    extra_env: dict[str, str] | None = None,
) -> subprocess.CompletedProcess[str]:
    try:
        return subprocess.run(
            command,
            cwd=cwd,
            env=clean_env(extra_env),
            text=True,
            encoding="utf-8",
            capture_output=True,
            timeout=timeout,
            check=False,
        )
    except subprocess.TimeoutExpired as exc:
        return subprocess.CompletedProcess(
            command,
            124,
            stdout=exc.stdout or "",
            stderr=(exc.stderr or "") + "command timed out\n",
        )


def copy_fixture(target: Path) -> None:
    shutil.copytree(FIXTURE, target, dirs_exist_ok=True, symlinks=True)


def initialize_git(target: Path) -> None:
    for command in (["git", "init", "--quiet"], ["git", "add", "."]):
        result = run(command, cwd=target, timeout=10)
        if result.returncode != 0:
            raise RuntimeError(f"{' '.join(command)} failed: {result.stderr}")


def validate(target: Path) -> subprocess.CompletedProcess[str]:
    return run([sys.executable, str(VALIDATOR), str(target)], cwd=target, timeout=30)


def main() -> int:
    failures: list[str] = []
    actual = sorted(
        path.relative_to(FIXTURE).as_posix()
        for path in FIXTURE.rglob("*")
        if not path.is_dir()
    )
    if actual != EXPECTED_FILES:
        failures.append(
            f"browser-interface: expected reduced layout {EXPECTED_FILES!r}, got {actual!r}"
        )

    runtime = (FIXTURE / "RUNTIME.md").read_text(encoding="utf-8")
    for command in (
        "TEXT_STATS_WEB_ENABLED=1 python web/server.py",
        "python web/server.py --stop",
        "python web/server.py --health",
        "python tests/test_web_server.py",
    ):
        if command not in runtime:
            failures.append(
                f"browser-interface runtime: missing documented command {command!r}"
            )

    with tempfile.TemporaryDirectory(prefix="browser-interface-profile") as temporary:
        target = Path(temporary) / "fixture"
        copy_fixture(target)
        initialize_git(target)

        validation = validate(target)
        if validation.returncode != 0:
            failures.append(
                "browser-interface: expected complete repository validation to pass; "
                f"stdout={validation.stdout!r}, stderr={validation.stderr!r}"
            )

        syntax = run(
            [
                sys.executable,
                "-m",
                "py_compile",
                "src/text_stats.py",
                "web/server.py",
                "tests/test_web_server.py",
            ],
            cwd=target,
            timeout=15,
        )
        if syntax.returncode != 0:
            failures.append(
                f"browser-interface syntax: stdout={syntax.stdout!r}, stderr={syntax.stderr!r}"
            )

        tests = run(
            [sys.executable, "tests/test_web_server.py"],
            cwd=target,
            timeout=90,
        )
        if tests.returncode != 0:
            failures.append(
                "browser-interface tests: expected success; "
                f"stdout={tests.stdout!r}, stderr={tests.stderr!r}"
            )

        implementation = target / "src/text_stats.py"
        missing = implementation.with_suffix(".py.missing")
        implementation.rename(missing)
        missing_implementation = run(
            [sys.executable, "web/server.py"],
            cwd=target,
            timeout=5,
            extra_env={
                "TEXT_STATS_WEB_ENABLED": "1",
                "TEXT_STATS_WEB_PORT": "0",
                "TEXT_STATS_WEB_PID_FILE": str(target / "missing.pid"),
            },
        )
        if missing_implementation.returncode == 0 or not missing_implementation.stderr.strip():
            failures.append(
                "browser-interface missing implementation: expected prompt nonzero failure with diagnostics; "
                f"status={missing_implementation.returncode!r}, "
                f"stderr={missing_implementation.stderr!r}"
            )
        missing.rename(implementation)

    with tempfile.TemporaryDirectory(prefix="invalid-browser-interface-profile") as temporary:
        target = Path(temporary)
        copy_fixture(target)
        (target / "WEB_INTERFACE.md").unlink()
        initialize_git(target)
        validation = validate(target)
        if validation.returncode == 0:
            failures.append(
                "browser-interface invalid contract: missing WEB_INTERFACE.md unexpectedly validates"
            )
        elif "WEB_INTERFACE.md" not in validation.stderr:
            failures.append(
                "browser-interface invalid contract: expected actionable WEB_INTERFACE.md diagnostic; "
                f"stderr={validation.stderr!r}"
            )

    if failures:
        for failure in failures:
            print(failure, file=sys.stderr)
        return 1
    print("Browser-interface profile fixture tests passed.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
